server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -572,4 +594,144 @@ public IndexMetadataTransferException(String errorDesc, Throwable cause) {
+            getBlobStoreTransferService().deleteAsync(
+                ThreadPool.Names.REMOTE_PURGE,
+                getCusterMetadataBasePath(clusterName, clusterUUID),
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void unused) {
+                        logger.info("Deleted all remote cluster metadata for cluster UUID - {}", clusterUUID);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.error(
+                            new ParameterizedMessage(
+                                "Exception occurred while deleting all remote cluster metadata for cluster UUID {}",
+                                clusterUUID
+                            ),
+                            e
+                        );
+                    }
+                }
+            );
+        });
+    }
+
+    /**
+     * Deletes older than last {@code versionsToRetain} manifests. Also cleans up unreferenced IndexMetadata associated with older manifests
+     * @param clusterName name of the cluster
+     * @param clusterUUID uuid of cluster state to refer to in remote
+     * @param manifestsToRetain no of latest manifest files to keep in remote
+     */
+    private void deleteStaleClusterMetadata(String clusterName, String clusterUUID, int manifestsToRetain) {
+        if (deleteStaleMetadataRunning.compareAndSet(false, true) == false) {
+            logger.info("Delete stale cluster metadata task is already in progress.");
+            return;
+        }
+        getBlobStoreTransferService().listAllInSortedOrderAsync(
+            ThreadPool.Names.REMOTE_PURGE,
+            getManifestFolderPath(clusterName, clusterUUID),
+            "manifest",
+            Integer.MAX_VALUE,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(List<BlobMetadata> blobMetadata) {
+                    if (blobMetadata.size() > manifestsToRetain) {
+                        List<ClusterMetadataManifest> allManifests = blobMetadata.stream()
+                            .map(
+                                mainfestBlobMetadata -> fetchRemoteClusterMetadataManifest(
+                                    clusterName,
+                                    clusterUUID,
+                                    mainfestBlobMetadata.name()
+                                )
+                            )
+                            .collect(Collectors.toList());
+                        deleteClusterMetadata(
+                            clusterName,
+                            clusterUUID,
+                            blobMetadata.subList(0, manifestsToRetain - 1),
+                            blobMetadata.subList(manifestsToRetain - 1, allManifests.size())
+                        );
+                    }
+                    deleteStaleMetadataRunning.set(false);
```

Comment:
It is important to always set this value to false even if there is an exception in the above `onResponse` code. We should add try-catch and have this in finally.

Reply:
entire `deleteClusterMetadata` is in try/catch. would that be anough or should we add a try/catch here as well?